### PR TITLE
Add @count util to torch, use it to track benchmark stats

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93013

image

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
This seems fine but:

1. It is duplicative with the existing compiler counters in Dynamo, and
2. The context manager is needlessly costly when you can do a less elegant direct increment call inside the body of the function you want to count, which will only be a bytecode

**Comment by voznesenskym** [comment]
> This seems fine but:
> 
> 1. It is duplicative with the existing compiler counters in Dynamo, and

These cant be trivially imported without a bunch of import shuffling, and there is no simple counter here as I could see

> 2. The context manager is needlessly costly when you can do a less elegant direct increment call inside the body of the function you want to count, which will only be a bytecode

This is not a context manager, just a decorator, its functionally equivalent to doing it within the body of the function, right?

**Review by ezyang — APPROVED** [review]

File: benchmarks/dynamo/common.py [hunk 1]
```diff
@@ -1518,9 +1518,14 @@ def run_one_model(
...
             print(status)
         if self.args.timing:
-            from torch._dynamo.utils import print_time_report
+            from torch._dynamo.utils import op_count, print_time_report
+            from torch.utils._stats import simple_call_counter
 
             print_time_report()
+            stats = f"STATS: call_* op count: {op_count}"
+            for key, value in simple_call_counter.items():
+                stats = f"{stats} | {key}:{value}"
```
Comment (ezyang) [inline benchmarks/dynamo/common.py:1527]: nit: `" | ".join(f"{key}:{value}" for key, value in simple_call_counter.items())`, mostly to get rid of the N^2 reinterpolation and it's about the same length of code nayway

Comment (voznesenskym) [reply benchmarks/dynamo/common.py:1527]: nice

File: torch/utils/_stats.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import collections
+import functools
+
+simple_call_counter = collections.OrderedDict()
+
+def count(fn):
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        if fn.__qualname__ not in simple_call_counter:
+            simple_call_counter[fn.__qualname__] = 0
```
Comment (ezyang) [inline torch/utils/_stats.py:13]: One downside to the QOL fix of avoiding having to type qualname is it makes it less easy to grep the codebase to answer the question "given this stat, what line of code am I actually measuring hits for"

File: benchmarks/dynamo/common.py (hunk 1 above)
Comment (ezyang) [reply benchmarks/dynamo/common.py:1527]: I know it doesn't matter, I just reflexively hate n^2 code
